cli/data.go

```diff
@@ -595,6 +595,113 @@ func (c *viamClient) dataAddToDataset(datasetID, orgID, locationID string, fileI
 	return nil
 }
 
+// DataAddToDatasetByFilter is the corresponding action for 'data dataset add filter'.
+func DataAddToDatasetByFilter(c *cli.Context) error {
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+	filter, err := createDataFilter(c)
+	if err != nil {
+		return err
+	}
+	if err := client.dataAddToDatasetByFilter(filter, c.String(datasetFlagDatasetID)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// dataAddToDatasetByFilter adds data, with the specified filter to the dataset corresponding to the dataset ID.
+func (c *viamClient) dataAddToDatasetByFilter(filter *datapb.Filter, datasetID string) error {
+	if err := c.ensureLoggedIn(); err != nil {
```

Comment:
Since this code is shared with `binaryData` method, if it's not too difficult, could you extract that into a common method that takes in some function to call below?

Here, that'd be `c.dataClient.AddBinaryDataToDatasetByIDs(context.Background(), &datapb.AddBinaryDataToDatasetByIDsRequest{DatasetId: datasetID, BinaryIds: []*datapb.BinaryID{id}})` and in `binaryData` that'd be `downloadBinary(ctx, c.dataClient, dst, id)`

Reply:
Refactored -- if this makes sense to you, will proceed in using in file download. Otherwise, I will revert.

While this does reduce some code overlap, I worry that the complexity of the code has increased a lot. Open to suggestions!

Reply:
I think this looks great and with the nit suggestion I had in the other comment, will be as simple and readable! Would just test that it still works with filter, IDs, and downloading binary data.